lib/scholar/interpolation/cubic_spline.ex

```diff
@@ -0,0 +1,189 @@
+defmodule Scholar.Interpolation.CubicSpline do
+  @moduledoc """
+  Cubic spline interpolation
+  """
+  import Nx.Defn
+
+  @derive {Nx.Container, containers: [:coefficients, :x]}
+  defstruct [:coefficients, :x]
+
+  opts = [
+    extrapolate: [
+      required: false,
+      default: true,
+      type: :boolean,
+      doc: "if false, out-of-bounds x return NaN."
+    ]
+  ]
+
+  @opts_schema NimbleOptions.new!(opts)
+
+  @doc """
+  Fits a cubic spline interpolation of the given `(x, y)` points
+
+  Inputs are expected to be rank-1 tensors with the same shape
+  and at least 3 entries.
+  """
+  defn train(x, y) do
+    # https://en.wikiversity.org/wiki/Cubic_Spline_Interpolation
+    # Reference implementation in Scipy
+
+    {n, dx} =
+      case Nx.shape(x) do
+        {n} when n > 2 ->
+          dx = x[1..-1//1] - x[0..-2//1]
```

Comment:
This doesn't need to be conditional, no?

Reply:
What do you mean? the guard?
It's because I haven't dealt with the case where only 2 points are given, so the code would also break

Reply:
I mean that we can compute `dx` separately, since it's not relevant for the check itself :)